plan/pattern.go

```diff
@@ -53,22 +63,28 @@ func Any() Pattern {
 	return &AnyPattern{}
 }
 
-// OneKindPattern matches a specified procedure with a predecessor pattern
+// UnionKindPattern matches a specified procedure with a predecessor pattern
 //
 //                   ProcedureKind
```

Comment:
This comment could be modified to better describe how this pattern works.

Reply:
Yeah I have not worked on a planner pattern matcher before so I wanted to know if there might be a better approach.

It has some similarities to a tree pattern matcher I worked on in a network security setting. I also maintained a `map[pattern root kind]  ->  set of rules`. In that case both the patterns and the thing they were tested against were trees and there was only one pass over the trees. Not fixed point because it didn't transform, was just looking for stuff that happened.

Reply:
Nice. In our case we're actually working with a DAG, so you might've noticed that we expicitly guard against matching patterns to nodes that have more than once successor. E.g., we won't be able to push the filter into the source for this case:
```
data = from(...) |> range(...)
data |> filter(...very selective filter...) |> yield(...)
data |> filter(...another very selective filter...) |> yield(...)
```
However, at some point, it would be nice to be able optimize this to something like:
```
from(...range and very selective filter pushed down...) |> yield(...)
from(...range and another very selective filter pushed down...) |> yield(...)
```
But I think we would need to start modeling cost in the planner to do that effectively. (I'm just mentioning it here so it's on your radar.)

Reply:
I was wondering about that ... is it feasible to "split apart" (for lack of a better expression) nodes such as `from` and `range` above when they are found to have multiple successors and it impedes a push down? I guess that's the optimization you mention. So what I'm really asking is ... can you just duplicate any node in the dag in order to satisfy a pushdown rule?

Reply:
With these pattern matching systems, is it normal to just make whatever pattern matching machinery you need in order to accomplish your task? Or is there a standard set of pattern forms that are normally made?

Reply:
> can you just duplicate any node in the dag in order to satisfy a pushdown rule?

As long as the semantics are preserved (e.g., the yield nodes produce the same result and we're not duplicating side-effecting nodes like `to`) and the new DAG is better by some metric, memory, CPU, etc, then I don't see why not. 

> With these pattern matching systems, is it normal to just make whatever pattern matching machinery you need in order to accomplish your task?

My sense is yes, that's normal.  I have only worked with pattern matchers that work on trees, so taking into account both predecessors and successors is new territory to me. I wonder if there is research on it.

The first paper (on the Volcano Optimizer) in this chapter of the Red Book might provide some useful background on query planning, if you're interested:
http://www.redbook.io/ch7-queryoptimization.html